mesonbuild/compilers/d.py

```diff
@@ -210,22 +213,12 @@ def translate_args_to_nongnu(cls, args):
             if arg.startswith('-Wl,'):
                 linkargs = arg[arg.index(',') + 1:].split(',')
                 for la in linkargs:
-                    dcargs.append('-L' + la.strip())
+                    dcargs.append(la.strip())
                 continue
             elif arg.startswith('-l'):
                 # translate library link flag
                 dcargs.append('-L' + arg)
                 continue
-            elif arg.startswith('-L/') or arg.startswith('-L./'):
-                # we need to handle cases where -L is set by e.g. a pkg-config
-                # setting to select a linker search path. We can however not
-                # unconditionally prefix '-L' with '-L' because the user might
-                # have set this flag too to do what it is intended to for this
-                # compiler (pass flag through to the linker)
-                # Hence, we guess here whether the flag was intended to pass
-                # a linker search path.
-                dcargs.append('-L' + arg)
```

Comment:
Removing the linker sanitization will wreck LDC and DMD support for sure, the only reason this does not result in an immediate CI failure is that the CI tests with gdc, not ldc or dmd.

The reason this is necessary is that even though the flags are linker flags, they are passed through to the linker via ldc or dmd which need to understand the commands properly (-L means something different unfortunately for the linker and for dmd/ldc).
Additionally, pkg-config files often contain additional linker flags that follow the GCC syntax, so if you remove this code you will break pkg-config support for a lot of C libraries as well.
This translation is actually so important that we even have a version of this code segment and the others in Debian itself to translate external build flags given in LDFLAGS/CFLAGS/DFLAGS.

So, to no surprise, I very much dislike the change, it will break a lot of existing code. Adjusting the new dub code seems to be a much safer option (the real fix would of course be to have ldc/dmd follow the GCC syntax like clang, but unfortunately that's very unlikely to happen).

Additionally to that, the commit making this change has no long commit message - with such an invasive change, I would expect at least some explanation given in the commit itself for why the code would no longer be necessary, ideally pointing at some regression testing.

Speaking of which, I wonder whether we should extend the CI to compile the D tests with LDC as well, that will make this issue more obvious to catch.

Reply:
Pretty much everything would break. I'll look into adding a few non-GDC testcases to Meson's CI somehow.
The current code has also been battle-tested in tens of thousands of builds of D code in Debian and on CIs, so I am certain it works the way it is.
Can you give me an example that didn't work for you with current Meson, what (Linux) system you use and which D compiler and whether you set the `DC` env var to something special?